File: problems/733/solution.cpp
class Solution {
public:
    vector<vector<int>> floodFill(vector<vector<int>>& image, int sr, int sc, int newColor) {
        queue<pair<int, int>> q;
        int colorToReplace = image[sr][sc],
            m = size(image),
            n = size(image.back());
        
        if (colorToReplace == newColor) {
            return image;
        }
        
        q.push({sr, sc});
        
        while (not empty(q)) {
            auto [x, y] = q.front();
            q.pop();
            
            if (x < 0 or y < 0 or x >= m or y >= n or image[x][y] != colorToReplace) {
                continue;
            }
            
            image[x][y] = newColor;
            
            q.push({x + 1, y});
            q.push({x - 1, y});
            q.push({x, y + 1});
            q.push({x, y - 1});
        }
        
        return image;
    }
};

File: problems/733/solution.py
class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, newColor: int) -> List[List[int]]:
        if image[sr][sc] == newColor:
            return image
        
        self.do(image, sr, sc, newColor, image[sr][sc])
        
        return image
    
    def do(self, img, x, y, newColor, oldColor):
        if x < 0 or x >= len(img) or y < 0 or y >= len(img[0]) or img[x][y] != oldColor:
            return
        
        img[x][y] = newColor
        
        self.do(img, x - 1, y, newColor, oldColor)
        self.do(img, x + 1, y, newColor, oldColor)
        self.do(img, x, y - 1, newColor, oldColor)
        self.do(img, x, y + 1, newColor, oldColor)